src/classes/TriggerHandler.cls

```diff
@@ -0,0 +1,238 @@
+    if(!validateRun()) return;
+
+    addToLoopCount();
+
+    // dispatch to the correct handler method
+    if(this.context == TriggerContext.BEFORE_INSERT) {
+      this.beforeInsert();
+    } else if(this.context == TriggerContext.BEFORE_UPDATE) {
+      this.beforeUpdate();
+    } else if(this.context == TriggerContext.BEFORE_DELETE) {
+      this.beforeDelete();
+    } else if(this.context == TriggerContext.AFTER_INSERT) {
+      this.afterInsert();
+    } else if(this.context == TriggerContext.AFTER_UPDATE) {
+      this.afterUpdate();
+    } else if(this.context == TriggerContext.AFTER_DELETE) {
+      this.afterDelete();
+    } else if(this.context == TriggerContext.AFTER_UNDELETE) {
+      this.afterUndelete();
+    }
+
+  }
+
+  public void setMaxLoopCount(Integer max) {
+    String handlerName = getHandlerName();
+    if(!TriggerHandler.loopCountMap.containsKey(handlerName)) {
+      TriggerHandler.loopCountMap.put(handlerName, new LoopCount(max));
+    } else {
+      TriggerHandler.loopCountMap.get(handlerName).setMax(max);
+    }
+  }
+
+  public void clearMaxLoopCount() {
+    this.setMaxLoopCount(-1);
+  }
+
+  /***************************************
+   * public static methods
+   ***************************************/
+
+  public static void bypass(String handlerName) {
+    TriggerHandler.bypassedHandlers.add(handlerName);
+  }
+
+  public static void clearBypass(String handlerName) {
+    TriggerHandler.bypassedHandlers.remove(handlerName);
+  }
+
+  public static Boolean isBypassed(String handlerName) {
+    return TriggerHandler.bypassedHandlers.contains(handlerName);
+  }
+
+  public static void clearAllBypasses() {
+    TriggerHandler.bypassedHandlers.clear();
+  }
+
+  /***************************************
+   * private instancemethods
+   ***************************************/
+
+  @TestVisible
+  private void setTriggerContext() {
+    this.setTriggerContext(null, false);
+  }
+
+  @TestVisible
+  private void setTriggerContext(String ctx, Boolean testMode) {
+    if(!Trigger.isExecuting && !testMode) {
+      this.isTriggerExecuting = false;
+      return;
+    } else {
+      this.isTriggerExecuting = true;
+    }
+    
+    if((Trigger.isExecuting && Trigger.isBefore && Trigger.isInsert) ||
+        (ctx != null && ctx == 'before insert')) {
+      this.context = TriggerContext.BEFORE_INSERT;
+    } else if((Trigger.isExecuting && Trigger.isBefore && Trigger.isUpdate) ||
+        (ctx != null && ctx == 'before update')){
+      this.context = TriggerContext.BEFORE_UPDATE;
+    } else if((Trigger.isExecuting && Trigger.isBefore && Trigger.isDelete) ||
+        (ctx != null && ctx == 'before delete')) {
+      this.context = TriggerContext.BEFORE_DELETE;
+    } else if((Trigger.isExecuting && Trigger.isAfter && Trigger.isInsert) ||
+        (ctx != null && ctx == 'after insert')) {
+      this.context = TriggerContext.AFTER_INSERT;
+    } else if((Trigger.isExecuting && Trigger.isAfter && Trigger.isUpdate) ||
```

Comment:
* * Classes and triggers should use the `UpperCamelCase` naming convention.
* Classes and triggers should use the `UpperCamelCase` naming convention.
* Classes and triggers should use the `UpperCamelCase` naming convention.